### DISCUSSION TITLE
How can one append text to page content if page uses a specific template?

### DISCUSSION THREAD

**Discussion by Andre601** [mkdocs/mkdocs#3863] [Q&A] [answered]
I would like to append some content to a page should it use a specific template and I would like to do this before any markdown processing, so that I can simply insert markdown and let the parser and extensions handle it.

In particular, I want to add the following at the bottom of a page's content, if it uses `blog-post.html` as its template:
```markdown
----

## :octicons-list-unordered-16: Footnotes { #__footnotes }

<!-- footnotes -->
```
Is this doable, or could there be an issue between the time of knowing the template used and parsing things?
I could also work with HTML itself if necessary (in fact I currently use HTML in my manual aproach using the snippets extension. Can't recall why I did that tbh and not used markdown...) tho I feel markdown is a bit cleaner in the end.

Anyways, would there be a way to do this with a hook?

**Comment by pawamoy** [comment]
What theme are you using? Does it support overriding templates?

**Reply by Andre601** [reply]
Yes.
I guess I could do that.

**Reply by Andre601** [reply]
Actually, looking at this, this would be a bit more complicated than anticipated.

I use Material for MkDocs and its Blog plugin, so the template in question I could customize is `blog-post.html`.
However, the main issue here is, that the page content isn't just added via `{{ content }}` but through a separate HTML template file (`partials/content.html`).

I can't just edit this partial, as it would affect all pages, unless there was a way to detect the child (or parent?) this partial is used in.
Tho, from testing, even if I did my own copy of this partial and use that, the `<!-- Footnotes -->` will be ignored here, because the markdown content - and by extension the... extensions - is handled separately from the template itself (Hope this makes sense).

**Comment by kamilkrzyskow** [comment, marked as answer]
Should be an easy hook, something like this,:
```py
def on_page_markdown(markdown: str, /, *, page, config, files) -> str | None:
    if page.meta["template"] != "blog-post.html":
        return
        
    return f"{markdown}\n{CONTENT}"
    
CONTENT: str = """
----

## :octicons-list-unordered-16: Footnotes { #__footnotes }

<!-- footnotes -->
"""
```

**Reply by Andre601** [reply]
Can give this a try.
Tho, not sure how important this is in Python, but shouldn't you also have `from mkdocs.structure.pages import Page` and have `page` changed to `page: Page` to have proper types?
Also, also, can't the other unused options (config and files) be ommited?

**Reply by Andre601** [reply]
Tested it and it works. Thanks.

**Reply by kamilkrzyskow** [reply]
In python types are optional, and mostly used for external tooling, auto complete etc..
To be fully correct, yes you should import it from the proper channels, but I didn't bother to check the correct import paths 😅 so just removed them after copying the definition from here:
- https://github.com/mkdocs/mkdocs/blob/7e4892ab2dd4a52efd95a9de407eee63310e0780/mkdocs/plugins.py#L345-L347


As for omitting the `config, files` arguments, I wouldn't recommend it, as there is the whole event loop that expects a certain "template" for the arguments of the called function. It might even work, but better to avoid edge cases.

You could do a `**_` to catch all the optional kwargs, under the `_` variable name, I even did such things in the past:
- https://github.com/Gothic-Modding-Community/gmc/blob/0a9b00648b89fb20653230c0aba28de18e03f8ae/overrides/.hooks/theme_overrides_manager.py#L72
but nowadays I rather stick to the actual template.
